### ISSUE TITLE
torch.compile changes fp16 NaN behavior for var + rsqrt normalization

### ISSUE THREAD

**Issue by ALinrunrun** [pytorch/pytorch#183605]
### 🐛 Describe the bug

`torch.compile` with the Inductor backend produces different fp16 behavior from eager execution for a variance-based normalization pattern.

For a small fp16 input, eager execution returns all NaNs, while the compiled version returns all zeros.

```
import torch

def fn(x):
    var = x.var(dim=-1, keepdim=True)
    return x * torch.rsqrt(var + 1e-9)

x = torch.full((3, 5), 1e-10, dtype=torch.float16)

eager = fn(x)

torch._dynamo.reset()
compiled = torch.compile(fn, backend="inductor")(x)

print("eager:", eager[0, 0].item(), "compile:", compiled[0, 0].item())
print(
    "eager_all_nan:",
    torch.isnan(eager).all().item(),
    "compile_all_zero:",
    (compiled == 0).all().item(),
)
```

### Error logs

eager: nan compile: 0.0
eager_all_nan: True compile_all_zero: True

### Versions

PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Laptop GPU
Nvidia driver version: 545.92
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] onnx==1.21.0
[pip3] onnx2torch==1.5.15
[pip3] onnxruntime==1.23.2
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by jansel** [comment]
My agent thinks this is indended:

This matches the documented Inductor default behavior in torch/_inductor/config.py for low precision fusion: eager fp16/bf16 pointwise execution inserts downcast/upcast boundaries, while Inductor default fusion may elide those boundaries and compute closer to fp64/reference numerics. The existing strict/eager numerics mode, TORCHINDUCTOR_EMULATE_PRECISION_CASTS=1, preserves the eager NaN behavior for this repro.
